File: backend/app/core/security.py
from datetime import datetime, timedelta, timezone
from typing import Any, Union
from jose import jwt
import bcrypt
from app.core.config import settings

def hash_password(password: str) -> str:
    salt = bcrypt.gensalt()
    hashed = bcrypt.hashpw(password.encode('utf-8'), salt)
    return hashed.decode('utf-8')

def verify_password(plain_password: str, hashed_password: str) -> bool:
    try:
        return bcrypt.checkpw(plain_password.encode('utf-8'), hashed_password.encode('utf-8'))
    except Exception:
        return False

def create_access_token(subject: Union[str, Any], role: str, expires_delta: timedelta = None) -> str:
    if expires_delta:
        expire = datetime.now(timezone.utc) + expires_delta
    else:
        expire = datetime.now(timezone.utc) + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode = {"exp": expire, "sub": str(subject), "role": role}
    return jwt.encode(to_encode, settings.JWT_SECRET, algorithm=settings.JWT_ALGORITHM)

def create_refresh_token(subject: Union[str, Any], role: str, expires_delta: timedelta = None) -> str:
    if expires_delta:
        expire = datetime.now(timezone.utc) + expires_delta
    else:
        expire = datetime.now(timezone.utc) + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
    to_encode = {"exp": expire, "sub": str(subject), "role": role}
    return jwt.encode(to_encode, settings.JWT_SECRET, algorithm=settings.JWT_ALGORITHM)

def decode_token(token: str) -> Union[dict, None]:
    try:
        return jwt.decode(token, settings.JWT_SECRET, algorithms=[settings.JWT_ALGORITHM])
    except jwt.JWTError:
        return None

File: backend/src/core/security.ts
import { randomUUID } from 'node:crypto';
import bcrypt from 'bcryptjs';
import { SignJWT, jwtVerify } from 'jose';
import { settings } from './config.js';

const secretKey = new TextEncoder().encode(settings.JWT_SECRET);

export type TokenKind = 'access' | 'refresh';

export interface TokenPayload {
  sub: string;
  role: string;
  /** Untyped legacy sessions must sign in again. */
  type: TokenKind;
  exp: number;
}

/**
 * bcryptjs là bản thuần JavaScript nên chậm hơn bcrypt native nhiều lần, và
 * chi phí tăng gấp đôi mỗi bậc: cost 12 trên VPS nhỏ mất vài giây cho một
 * lần đăng ký. Cost 10 (mặc định của chính bcrypt, và mức tối thiểu OWASP
 * khuyến nghị) nhanh hơn bốn lần.
 *
 * Đổi con số này không làm hỏng tài khoản cũ: bcrypt nhét cost vào trong
 * chuỗi hash, nên hash $2b$12$... vẫn kiểm tra đúng như thường.
 */
const BCRYPT_ROUNDS = Number(process.env.BCRYPT_ROUNDS ?? 10);

export async function hashPassword(password: string): Promise<string> {
  return bcrypt.hash(password, BCRYPT_ROUNDS);
}

/**
 * Verifies against the bcrypt hashes written by the Python stack. The stored
 * `$2b$` digests must keep working, so the algorithm cannot be swapped.
 */
export async function verifyPassword(plain: string, hashed: string): Promise<boolean> {
  try {
    return await bcrypt.compare(plain, hashed);
  } catch {
    return false;
  }
}

async function sign(sub: string, role: string, type: TokenKind, expiresIn: string) {
  return new SignJWT({ role, type })
    .setProtectedHeader({ alg: settings.JWT_ALGORITHM })
    .setSubject(String(sub))
    .setIssuedAt()
    // Refreshes in the same second must still issue distinct credentials.
    .setJti(randomUUID())
    .setExpirationTime(expiresIn)
    .sign(secretKey);
}

export function createAccessToken(sub: string, role: string) {
  return sign(sub, role, 'access', `${settings.ACCESS_TOKEN_EXPIRE_MINUTES}m`);
}

export function createRefreshToken(sub: string, role: string) {
  return sign(sub, role, 'refresh', `${settings.REFRESH_TOKEN_EXPIRE_DAYS}d`);
}

/** Verify the kind, expiry and UUID before any user lookup; invalid tokens return null. */
export async function decodeToken(token: string, expectedType: TokenKind): Promise<TokenPayload | null> {
  try {
    const { payload } = await jwtVerify(token, secretKey, {
      algorithms: [settings.JWT_ALGORITHM],
      requiredClaims: ['sub', 'exp', 'type'],
    });
    const sub = asUuid(payload.sub);
    if (!sub || payload.type !== expectedType || typeof payload.exp !== 'number') return null;
    return {
      sub,
      role: typeof payload.role === 'string' ? payload.role : '',
      type: expectedType,
      exp: payload.exp,
    };
  } catch {
    return null;
  }
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

/**
 * JWT subjects come back as strings; uuid columns need a real UUID.
 * Postgres raises on a malformed uuid literal, so every id that reaches a
 * query goes through here first (auth.py:54 skipped this and crashed).
 */
export function asUuid(value: unknown): string | null {
  if (typeof value !== 'string') return null;
  return UUID_RE.test(value) ? value : null;
}
